pip/pip-331.md

```diff
@@ -0,0 +1,129 @@
+# PIP-331: WASM Function API
+
+# Background knowledge
+
+WASM(WebAssembly) bytecode is designed to be encoded in a size- and load-time-efficient binary format. WASM aims to leverage the common hardware features available on various platforms to execute in browsers at machine code speed.
+
+WASI(WebAssembly System Interface) provide a portable interface for applications that run within a constrained sandbox environment, which allows WASM to run in non browser environments such as Linux. It's portable and secure.
+
+# Motivation
+
+The server and client sides of the Pulsar function use protobuf for decoupling. In principle, the language supported by protobuf can be supported by the pulsar function, now Pulsar provided the java, python and golang function client, but there are still many languages that are not supported.
+
+Before all language adaptations are completed (and it's almost entirely certain to be impossible), users cannot write pulsar function in their familiar languages.
+
+# Goals
+
+## In Scope
+
+Other languages, as long as their code can be compiled into WASM bytecode (such as Rust/golang/C++), users can use these languages to write pulsar function.
+
+## Out of Scope
+
+All existing abilities of the Java pulsar function client are not reimplemented, the WASM Pulsar functions is under the Java Pulsar functions.
+
+Due to the strict requirements of WASM on parameter types and for simplicity reasons, types other than `java.lang.Long` are not used as parameters or return value.
```

Comment:
In brief, Java passes parameter IDs(`java.lang.Long`) to the other language, and then the other language obtain the raw parameters through the parameter ID, the raw parameters can be topics, messages, and so on.

The more details in https://github.com/apache/pulsar/pull/21975#issuecomment-1928993883

Reply:
I think there is a problem here. I can't learn about this feature you mentioned from the documentation of wasmtime-java since it's non-existent, and the scarce documentation in this PIP doesn't help. How can this be maintained by other people going forward?

Reply:
Sorry, I will update the document this weekend.